File: Classes/Model classes/SharedWatch.m
/*
	MyFlightbook for iOS - provides native access to MyFlightbook
	pilot's logbook
 Copyright (C) 2017-2020 MyFlightbook, LLC
 
 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

//
//  SharedWatch.m
//  MFBSample
//
//  Created by Eric Berman on 10/29/15.
//
//

#import "SharedWatch.h"

#define keyWatchLatitude @"LAT"
#define keyWatchLongitude @"LON"
#define keyWatchSpeed @"SPEED"
#define keyWatchAlt @"ALT"
#define keyWatchFlightStatus @"FLIGHTSTATUS"
#define keyWatchElapsed @"FLIGHTELAPSED"
#define keyWatchIsPaused @"FLIGHTPAUSED"
#define keyWatchIsRecording @"FLIGHTRECORDING"
#define keyWatchFlightStage @"FlightStage"
#define keyWatchLatestFlight @"LASTFLIGHT"

@implementation SharedWatch

@synthesize latDisplay, lonDisplay, speedDisplay, altDisplay, flightstatus, isPaused, isRecording, elapsedSeconds, flightStage;

- (instancetype) init {
    if (self = [super init])
    {
        self.latDisplay = self.lonDisplay = self.speedDisplay = self.altDisplay = self.flightstatus = @"";
        self.isPaused = self.isRecording = NO;
        self.flightStage = flightStageUnknown;
        self.latestFlight = nil;
    }
    return self;
}

-(void) dealloc{
    self.latDisplay = self.lonDisplay = self.speedDisplay = self.altDisplay = self.flightstatus = nil;
}

+ (BOOL)supportsSecureCoding {return YES;}

- (instancetype) initWithCoder:(NSCoder *)aDecoder
{
    self.latDisplay = [aDecoder decodeObjectForKey:keyWatchLatitude];
    self.lonDisplay = [aDecoder decodeObjectForKey:keyWatchLongitude];
    self.altDisplay = [aDecoder decodeObjectForKey:keyWatchAlt];
    self.speedDisplay = [aDecoder decodeObjectForKey:keyWatchSpeed];
    self.flightstatus = [aDecoder decodeObjectForKey:keyWatchFlightStatus];
    self.latestFlight = [aDecoder decodeObjectForKey:keyWatchLatestFlight];
    self.isPaused = [aDecoder decodeBoolForKey:keyWatchIsPaused];
    self.isRecording = [aDecoder decodeBoolForKey:keyWatchIsRecording];
    self.elapsedSeconds = [aDecoder decodeDoubleForKey:keyWatchElapsed];
    self.flightStage = [aDecoder decodeIntegerForKey:keyWatchFlightStage];
    return self;
}

- (void) encodeWithCoder:(NSCoder *)aCoder
{
    [aCoder encodeObject:self.latDisplay forKey:keyWatchLatitude];
    [aCoder encodeObject:self.lonDisplay forKey:keyWatchLongitude];
    [aCoder encodeObject:self.altDisplay forKey:keyWatchAlt];
    [aCoder encodeObject:self.speedDisplay forKey:keyWatchSpeed];
    [aCoder encodeObject:self.flightstatus forKey:keyWatchFlightStatus];
    [aCoder encodeObject:self.latestFlight forKey:keyWatchLatestFlight];
    [aCoder encodeBool:self.isPaused forKey:keyWatchIsPaused];
    [aCoder encodeBool:self.isRecording forKey:keyWatchIsRecording];
    [aCoder encodeDouble:self.elapsedSeconds forKey:keyWatchElapsed];
    [aCoder encodeInteger:self.flightStage forKey:keyWatchFlightStage];
}
@end

#pragma mark SimpleCurrencyItem
#define keyCurAttribute @"curAttribute"
#define keyCurValue @"curValue"
#define keyCurDiscrepancy @"curDiscrepancy"
#define keyCurState @"curState"

@implementation SimpleCurrencyItem

@synthesize attribute, value, discrepancy, state;

- (instancetype) init {
    if (self = [super init])
    {
        self.attribute = self.value = self.discrepancy = @"";
        self.state = MFBWebServiceSvc_CurrencyState_none;
    }
    return self;
}

+ (BOOL)supportsSecureCoding {return YES;}

- (void) encodeWithCoder:(NSCoder *)aCoder
{
    [aCoder encodeObject:self.attribute forKey:keyCurAttribute];
    [aCoder encodeObject:self.value forKey:keyCurValue];
    [aCoder encodeObject:self.discrepancy forKey:keyCurDiscrepancy];
    [aCoder encodeInteger:self.state forKey:keyCurState];
}

- (instancetype) initWithCoder:(NSCoder *)aDecoder
{
    self.attribute = [aDecoder decodeObjectForKey:keyCurAttribute];
    self.value = [aDecoder decodeObjectForKey:keyCurValue];
    self.discrepancy = [aDecoder decodeObjectForKey:keyCurDiscrepancy];
    self.state =  (MFBWebServiceSvc_CurrencyState) [aDecoder decodeIntegerForKey:keyCurState];
    return self;
}

@end

#pragma mark SimpleTotalItem

#define keyTotalTitle @"totTitle"
#define keyTotalValue @"totValue"
#define keyTotalSubDesc @"totSubDesc"

@implementation SimpleTotalItem

@synthesize title, valueDisplay, subDesc;

- (instancetype) init {
    if (self = [super init])
    {
        self.title = self.valueDisplay = self.subDesc = @"";
    }
    return self;
}

- (void) encodeWithCoder:(NSCoder *)aCoder
{
    [aCoder encodeObject:self.title forKey:keyTotalTitle];
    [aCoder encodeObject:self.valueDisplay forKey:keyTotalValue];
    [aCoder encodeObject:self.subDesc forKey:keyTotalSubDesc];
}

+ (BOOL)supportsSecureCoding {return YES;}

- (instancetype) initWithCoder:(NSCoder *)aDecoder
{
    self.title = [aDecoder decodeObjectForKey:keyTotalTitle];
    self.valueDisplay = [aDecoder decodeObjectForKey:keyTotalValue];
    self.subDesc = [aDecoder decodeObjectForKey:keyTotalSubDesc];
    return self;
}

@end

#pragma mark SimpleLogbookEntry

#define keyLEComment @"LEComment"
#define keyLERoute @"LERoute"
#define keyLEDate @"LEDate"
#define keyLETotal @"LETotal"
#define keyLETailDisplay @"LETail"

@implementation SimpleLogbookEntry

@synthesize Comment, Route, Date, TotalTimeDisplay, TailNumDisplay;

- (instancetype) init {
    if (self = [super init]) {
        self.Comment = self.Route = self.TotalTimeDisplay = @"";
        self.Date = [NSDate date];
    }
    return self;
}

+ (BOOL)supportsSecureCoding {return YES;}


- (void) encodeWithCoder:(NSCoder *)aCoder {
    [aCoder encodeObject:self.Comment forKey:keyLEComment];
    [aCoder encodeObject:self.Route forKey:keyLERoute];
    [aCoder encodeObject:self.Date forKey:keyLEDate];
    [aCoder encodeObject:self.TotalTimeDisplay forKey:keyLETotal];
    [aCoder encodeObject:self.TailNumDisplay forKey:keyLETailDisplay];
}

- (instancetype) initWithCoder:(NSCoder *)aDecoder {
    self.Comment = [aDecoder decodeObjectForKey:keyLEComment];
    self.Route = [aDecoder decodeObjectForKey:keyLERoute];
    self.Date = [aDecoder decodeObjectForKey:keyLEDate];
    self.TotalTimeDisplay = [aDecoder decodeObjectForKey:keyLETotal];
    self.TailNumDisplay = [aDecoder decodeObjectForKey:keyLETailDisplay];
    return self;
}
@end

File: Classes/SharedWatch.swift
/*
    MyFlightbook for iOS - provides native access to MyFlightbook
    pilot's logbook
 Copyright (C) 2023 MyFlightbook, LLC
 
 This program is free software: you can redistribute it and/or modify
 it under the terms of the GNU General Public License as published by
 the Free Software Foundation, either version 3 of the License, or
 (at your option) any later version.
 
 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
//
//  SharedWatch.swift
//  MyFlightbook
//
//  Created by Eric Berman on 3/22/23.
//

import Foundation

@objc public enum NewFlightStages : Int {
    case unknown
    case unstarted
    case inprogress
    case done
}

// Dictionary key for kinds of messages
public let WATCH_MESSAGE_REQUEST_DATA = "messageRequestData"
public let WATCH_MESSAGE_ACTION = "messageRequestAction"

// Dictionary key for data requests
public let WATCH_REQUEST_STATUS = "watchRequestStatus"
public let WATCH_REQUEST_CURRENCY = "watchRequestCurrency"
public let WATCH_REQUEST_TOTALS = "watchRequestTotals"
public let WATCH_REQUEST_RECENTS = "watchRequestRecents"
public let WATCH_REQUEST_GLANCE = "watchRequestGlance"

// Dictionary key for result data
public let WATCH_RESPONSE_STATUS  = "sharedwatchStatus"
public let WATCH_RESPONSE_CURRENCY = "sharedwatchCurrency"
public let WATCH_RESPONSE_TOTALS = "sharedwatchTotals"
public let WATCH_RESPONSE_RECENTS = "sharedWatchRecents"

// Dictionary key for requested actions
public let WATCH_ACTION_START = "actionStartFlight"
public let WATCH_ACTION_END = "actionEndFlight"
public let WATCH_ACTION_TOGGLE_PAUSE = "actionTogglePause"

// Note that we MUST decorate the classes with "@objc(name-of-class)" so that these are correctly received from a namespace perspective by the watchkit extension.
// Otherwise archive/dearchive doesn't work because the class names (with namespaces) don't line up.
// Alternatively, we could also put these into a framework
// See https://stackoverflow.com/questions/29472935/cannot-decode-object-of-class for more information.

@objc(SharedWatch) public class SharedWatch : NSObject, NSCoding, NSSecureCoding {
    private let keyWatchLatitude = "LAT"
    private let keyWatchLongitude = "LON"
    private let keyWatchSpeed = "SPEED"
    private let keyWatchAlt = "ALT"
    private let keyWatchFlightStatus = "FLIGHTSTATUS"
    private let keyWatchElapsed = "FLIGHTELAPSED"
    private let keyWatchIsPaused = "FLIGHTPAUSED"
    private let keyWatchIsRecording = "FLIGHTRECORDING"
    private let keyWatchFlightStage = "FlightStage"
    private let keyWatchLatestFlight = "LASTFLIGHT"
    
    @objc public var latDisplay = ""
    @objc public var lonDisplay = ""
    @objc public var speedDisplay = ""
    @objc public var altDisplay = ""
    @objc public var flightstatus = ""
    @objc public var isPaused = false
    @objc public var isRecording = false
    @objc public var elapsedSeconds = 0.0
    @objc public var flightStage = NewFlightStages.unknown
    @objc public var latestFlight : SimpleLogbookEntry? = nil
    
    @objc public override init() {
        super.init()
    }

    @objc public required init?(coder: NSCoder) {
        latDisplay = coder.decodeObject(forKey: keyWatchLatitude) as? String ?? ""
        lonDisplay = coder.decodeObject(forKey: keyWatchLongitude) as? String ?? ""
        altDisplay = coder.decodeObject(forKey: keyWatchAlt) as? String ?? ""
        speedDisplay = coder.decodeObject(forKey: keyWatchSpeed) as? String ?? ""
        flightstatus = coder.decodeObject(forKey: keyWatchFlightStatus) as? String ?? ""
        latestFlight = coder.decodeObject(forKey: keyWatchLatestFlight) as? SimpleLogbookEntry
        isPaused = coder.decodeBool(forKey: keyWatchIsPaused)
        isRecording = coder.decodeBool(forKey: keyWatchIsRecording)
        elapsedSeconds = coder.decodeDouble(forKey: keyWatchElapsed)
        flightStage = NewFlightStages(rawValue: coder.decodeInteger(forKey: keyWatchFlightStage))!
    }
    
    @objc public func encode(with coder: NSCoder) {
        coder.encode(latDisplay, forKey: keyWatchLatitude)
        coder.encode(lonDisplay, forKey: keyWatchLongitude)
        coder.encode(altDisplay, forKey: keyWatchAlt)
        coder.encode(speedDisplay, forKey: keyWatchSpeed)
        coder.encode(flightstatus, forKey: keyWatchFlightStatus)
        coder.encode(latestFlight, forKey: keyWatchLatestFlight)
        coder.encode(isPaused, forKey: keyWatchIsPaused)
        coder.encode(isRecording, forKey: keyWatchIsRecording)
        coder.encode(elapsedSeconds, forKey: keyWatchElapsed)
        coder.encode(flightStage.rawValue, forKey: keyWatchFlightStage)
    }
    
    @objc public static var supportsSecureCoding: Bool {
        return true
    }
}
